### PR TITLE
MES-2353: Retry Processing Lambda

### PR DESCRIPTION
**Pull request by rossfauldsbjss**
MES 2353 - lambda to retry test result/upload queues. 

This lambda updates database state to set stalled/failed/ completed jobs to a state that allows other lambdas to process them.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,16 +14,19 @@
     "@types/hapi__joi": "^15.0.1",
     "enjoi": "^6.0.0",
     "moment": "^2.23.0",
-    "mysql2": "^1.6.5"
+    "mysql2": "^1.6.5",
+    "sqlite3": "^4.0.9"
```
Comment (rstepnowski) [inline package.json]: We use this only in tests -> should this be in devDeps?

File: src/functions/retryProcessing/application/RetryProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      const [rows] = await connection.promise().query(buildSuccessfullyProcessedQuery());
+
+      for (const row of rows) {
+        const [updated] = await connection.promise().query(buildUpdateTestResultStatusQuery(
+          row.application_reference,
+          row.staff_number,
+          'PROCESSED',
+        ));
+      }
+    } catch (err) {
```
Comment (rstepnowski) [inline src/functions/retryProcessing/application/RetryProcessor.ts]: More detailed log messages on error scenarios would be useful here (for all the queries)

File: src/functions/retryProcessing/application/RetryProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      retryConfig().notifyRetryCount,
+      retryConfig().tarsRetryCount,
+    );
+
+    await this.processSupportInterventions();
+    await this.processOldEntryCleanup(retryConfig().retryCutOffPointDays);
+  }
+
+  async processSuccessful(): Promise<void> {
+    const connection: mysql.Connection = getConnection();
```
Comment (rstepnowski) [inline src/functions/retryProcessing/application/RetryProcessor.ts]: We probably don't need every query to open a new connection...

File: src/functions/retryProcessing/framework/handler.ts [hunk 4]
```diff
@@ -1,8 +1,20 @@
...
+import { bootstrapConfig } from '../../../../src/common/framework/config/config';
+import { IRetryProcessor } from '../application/IRetryProcessor';
+import { error } from '@dvsa/mes-microservice-common/application/utils/logger';
 
-export async function handler(event: APIGatewayEvent, fnCtx: Context): Promise<Response> {
+export async function handler(event: ScheduledEvent, fnCtx: Context): Promise<Response> {
+  await bootstrapConfig();
+  const retryProcessor: IRetryProcessor = new RetryProcessor();
+  try {
+    await retryProcessor.processRetries();
```
Comment (rstepnowski) [inline src/functions/retryProcessing/framework/handler.ts]: We would like to continue processing the rest of the queries even though one of them failed (got rollbacked)

File: package.json (hunk 1 above)
Comment (johnmorleybjss) [reply package.json]: Good spot, moved to `devDependencies`.

File: src/functions/retryProcessing/application/RetryProcessor.ts (hunk 2 above)
Comment (johnmorleybjss) [reply src/functions/retryProcessing/application/RetryProcessor.ts]: I have added a warning message to each catch in `RetryProcessor.ts` using the microservice common logger.

File: src/functions/retryProcessing/application/RetryProcessor.ts (hunk 3 above)
Comment (johnmorleybjss) [reply src/functions/retryProcessing/application/RetryProcessor.ts]: `getConnection()` and `connection.end()` are now only called in the handler.

File: src/functions/retryProcessing/framework/handler.ts (hunk 4 above)
Comment (johnmorleybjss) [reply src/functions/retryProcessing/framework/handler.ts]: Removed the throw on each each in favour of logging.

**Pull request merged**